<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="prev" title="API reference" href="modules.html" />

    <!-- Generated with Sphinx 6.1.3 and Furo 2022.12.07 -->
        <title>phyloformer package - Phyloformer 0.0.1a3 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?digest=91d0f0d1c444bdcb17a68e833c7a53903343c195" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">Phyloformer 0.0.1a3 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">Phyloformer 0.0.1a3 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="readme.html">Introduction to Phyloformer</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="modules.html">API reference</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">phyloformer package</a></li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="phyloformer-package">
<h1>phyloformer package<a class="headerlink" href="#phyloformer-package" title="Permalink to this heading">#</a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this heading">#</a></h2>
</section>
<section id="module-phyloformer.attentions">
<span id="phyloformer-attentions-module"></span><h2>phyloformer.attentions module<a class="headerlink" href="#module-phyloformer.attentions" title="Permalink to this heading">#</a></h2>
<p>The attentions module contains classes for the MultiHeadAttention modules
in the Phyloformer network</p>
<dl class="py class">
<dt class="sig sig-object py" id="phyloformer.attentions.KernelAxialMultiAttention">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">phyloformer.attentions.</span></span><span class="sig-name descname"><span class="pre">KernelAxialMultiAttention</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">h_dim</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_heads</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dropout</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-06</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/phyloformer/attentions.html#KernelAxialMultiAttention"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.attentions.KernelAxialMultiAttention" title="Permalink to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Module</span></code></p>
<p>Initializes internal Module state, shared by both nn.Module and ScriptModule.</p>
<dl class="py method">
<dt class="sig sig-object py" id="phyloformer.attentions.KernelAxialMultiAttention.forward">
<span class="sig-name descname"><span class="pre">forward</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/phyloformer/attentions.html#KernelAxialMultiAttention.forward"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.attentions.KernelAxialMultiAttention.forward" title="Permalink to this definition">#</a></dt>
<dd><p>Defines the computation performed at every call.</p>
<p>Should be overridden by all subclasses.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Although the recipe for forward pass needs to be defined within
this function, one should call the <code class="xref py py-class docutils literal notranslate"><span class="pre">Module</span></code> instance afterwards
instead of this since the former takes care of running the
registered hooks while the latter silently ignores them.</p>
</div>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="phyloformer.attentions.KernelAxialMultiAttention.training">
<span class="sig-name descname"><span class="pre">training</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">bool</span></em><a class="headerlink" href="#phyloformer.attentions.KernelAxialMultiAttention.training" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

</dd></dl>

</section>
<section id="module-phyloformer.data">
<span id="phyloformer-data-module"></span><h2>phyloformer.data module<a class="headerlink" href="#module-phyloformer.data" title="Permalink to this heading">#</a></h2>
<p>The data module contains functions that are used to load alignment data in a format that
the Phyloformer network understands.</p>
<dl class="py class">
<dt class="sig sig-object py" id="phyloformer.data.TensorDataset">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">phyloformer.data.</span></span><span class="sig-name descname"><span class="pre">TensorDataset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">directory</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/phyloformer/data.html#TensorDataset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.data.TensorDataset" title="Permalink to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Dataset</span></code></p>
<p>A Dataset class to train Phyloformer networks</p>
<p>Instanciates a TensorDataset</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>directory</strong> (<em>str</em>) – Path to the directory containing .tensor_pair files generated by the
<cite>make_tensors</cite> script.</p></li>
<li><p><strong>filter</strong> (<em>List</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – List of tensor pair names to keep (useful if you keep training and
validation tensors in the same directory), default is None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A instance of TensorDataset for training phyloformer</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#phyloformer.data.TensorDataset" title="phyloformer.data.TensorDataset">TensorDataset</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="phyloformer.data.load_alignment">
<span class="sig-prename descclassname"><span class="pre">phyloformer.data.</span></span><span class="sig-name descname"><span class="pre">load_alignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">Tensor</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/phyloformer/data.html#load_alignment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.data.load_alignment" title="Permalink to this definition">#</a></dt>
<dd><p>Loads an alignment into a tensor digestible by the Phyloformer network</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>path</strong> (<em>str</em>) – Path to a fasta file containing the alignment</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>a tuple containing:</dt><dd><ul class="simple">
<li><p>a tensor representing the alignment (shape 22 * seq_len * n_seq)</p></li>
<li><p>a list of ids of the sequences in the alignment</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[torch.Tensor, List[str]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="phyloformer.data.load_dataset">
<span class="sig-prename descclassname"><span class="pre">phyloformer.data.</span></span><span class="sig-name descname"><span class="pre">load_dataset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">Tensor</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/phyloformer/data.html#load_dataset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.data.load_dataset" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="phyloformer.data.load_tree">
<span class="sig-prename descclassname"><span class="pre">phyloformer.data.</span></span><span class="sig-name descname"><span class="pre">load_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">Tensor</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/phyloformer/data.html#load_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.data.load_tree" title="Permalink to this definition">#</a></dt>
<dd><p>Loads a tree as a tensor of pairwise distances, digestible by the Phyloformer
network</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>path</strong> (<em>str</em>) – Path to the newick file containing the tree</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>a tuple containing:</dt><dd><ul class="simple">
<li><p>a tensor representing the distance matrix of the tree (shape 1*n_pairs)</p></li>
<li><p>a list of tuples of ids indicating between which leafs the distance was
computed</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[torch.Tensor, List[Tuple[str, str]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="phyloformer.data.write_dm">
<span class="sig-prename descclassname"><span class="pre">phyloformer.data.</span></span><span class="sig-name descname"><span class="pre">write_dm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dm</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DistanceMatrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/phyloformer/data.html#write_dm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.data.write_dm" title="Permalink to this definition">#</a></dt>
<dd><p>Write a distance matrix to disk in the square Phylip matrix format</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dm</strong> (<em>skbio.DistanceMatrix</em>) – Distance matrix to save</p></li>
<li><p><strong>path</strong> (<em>str</em>) – Path where to save the matrix</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-phyloformer.phyloformer">
<span id="phyloformer-phyloformer-module"></span><h2>phyloformer.phyloformer module<a class="headerlink" href="#module-phyloformer.phyloformer" title="Permalink to this heading">#</a></h2>
<p>The phyloformer module contains the Phyloformer network as well as functions to
create and load instances of the network from disk</p>
<dl class="py class">
<dt class="sig sig-object py" id="phyloformer.phyloformer.AttentionNet">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">phyloformer.phyloformer.</span></span><span class="sig-name descname"><span class="pre">AttentionNet</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_blocks</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_heads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">h_dim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">64</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dropout</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'cpu'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_seqs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_len</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/phyloformer/phyloformer.html#AttentionNet"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.phyloformer.AttentionNet" title="Permalink to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Module</span></code></p>
<p>Phyloformer Network</p>
<p>Initializes internal Module state</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n_blocks</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of blocks in transformer, by default 1</p></li>
<li><p><strong>n_heads</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of heads in multi-head attention, by default 4</p></li>
<li><p><strong>h_dim</strong> (<em>int</em><em>, </em><em>optional</em>) – Hidden dimension, by default 64</p></li>
<li><p><strong>dropout</strong> (<em>float</em><em>, </em><em>optional</em>) – Droupout rate, by default 0.0</p></li>
<li><p><strong>device</strong> (<em>str</em><em>, </em><em>optional</em>) – Device for model (“cuda” or “cpu”), by default “cpu”</p></li>
<li><p><strong>n_seqs</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of sequences in input alignments, by default 20</p></li>
<li><p><strong>seq_len</strong> (<em>int</em><em>, </em><em>optional</em>) – Length of sequences in input alignment, by default 200</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Functional instance of AttentionNet for inference/fine-tuning</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet">AttentionNet</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If h_dim is not divisible by n_heads</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="phyloformer.phyloformer.AttentionNet.forward">
<span class="sig-name descname"><span class="pre">forward</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Tensor</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tensor</span></span></span><a class="reference internal" href="_modules/phyloformer/phyloformer.html#AttentionNet.forward"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.phyloformer.AttentionNet.forward" title="Permalink to this definition">#</a></dt>
<dd><p>Doed a forward pass through the Phyloformer network</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>x</strong> (<em>torch.Tensor</em>) – Input tensor (shape 1*22*n_seqs*seq_len)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Output tensor (shape 1*n_pairs)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>torch.Tensor</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the tensors aren’t the right shape</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="phyloformer.phyloformer.AttentionNet.infer_dm">
<span class="sig-name descname"><span class="pre">infer_dm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Tensor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DistanceMatrix</span></span></span><a class="reference internal" href="_modules/phyloformer/phyloformer.html#AttentionNet.infer_dm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.phyloformer.AttentionNet.infer_dm" title="Permalink to this definition">#</a></dt>
<dd><p>Infers a phylogenetic distance matrix from embedded alignment tensor</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>torch.Tensor</em>) – Input alignment, embedded as a tensor (shape 22*n_seq*seq_len)</p></li>
<li><p><strong>ids</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Identifiers of the sequences in the input tensor, by default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Phylolgenetic distance matrix inferred by Phyloformer</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>skbio.DistanceMatrix</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the tensors aren’t the right shape</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="phyloformer.phyloformer.AttentionNet.infer_tree">
<span class="sig-name descname"><span class="pre">infer_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Tensor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dm</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DistanceMatrix</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">TreeNode</span></span></span><a class="reference internal" href="_modules/phyloformer/phyloformer.html#AttentionNet.infer_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.phyloformer.AttentionNet.infer_tree" title="Permalink to this definition">#</a></dt>
<dd><p>Infers a phylogenetic tree from an embedded alignment tensor</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>torch.Tensor</em>) – Input alignment, embedded as a tensor (shape 22*n_seq*seq_len)</p></li>
<li><p><strong>ids</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Identifiers of the sequences in the input tensor, by default None</p></li>
<li><p><strong>dm</strong> (<em>skbio.DistanceMatrix</em><em>, </em><em>optional</em>) – Precomputed distance matrix if you have already run <cite>AttentionNet.infer_dm</cite>
on your own, by default None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Phylogenetic tree computed with neighbour joining from the distance matrix
inferred by Phyloformer</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tree</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the tensors aren’t the right shape</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="phyloformer.phyloformer.AttentionNet.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/phyloformer/phyloformer.html#AttentionNet.save"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.phyloformer.AttentionNet.save" title="Permalink to this definition">#</a></dt>
<dd><p>Saves the model parameters to disk</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>path</strong> (<em>str</em>) – Path to save the model to</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="phyloformer.phyloformer.AttentionNet.training">
<span class="sig-name descname"><span class="pre">training</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">bool</span></em><a class="headerlink" href="#phyloformer.phyloformer.AttentionNet.training" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="phyloformer.phyloformer.load_model">
<span class="sig-prename descclassname"><span class="pre">phyloformer.phyloformer.</span></span><span class="sig-name descname"><span class="pre">load_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">single_gpu</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet"><span class="pre">AttentionNet</span></a></span></span><a class="reference internal" href="_modules/phyloformer/phyloformer.html#load_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.phyloformer.load_model" title="Permalink to this definition">#</a></dt>
<dd><p>Load a Phyloformer istance froms disk</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<em>str</em>) – Path to model saved with AttentionNet.save()</p></li>
<li><p><strong>single_gpu</strong> (<em>bool</em><em>, </em><em>optional</em>) – Wether inference/fine-tuning will be done on a single GPU, by default True</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Functional instance of AttentionNet for inference/fine-tuning</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet">AttentionNet</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the file does not contain the state_dict and model architecture parameters</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="phyloformer.phyloformer.load_state_dict">
<span class="sig-prename descclassname"><span class="pre">phyloformer.phyloformer.</span></span><span class="sig-name descname"><span class="pre">load_state_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_blocks</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">h_dim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">64</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_heads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_seqs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_len</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'cpu'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dropout</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">single_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet"><span class="pre">AttentionNet</span></a></span></span><a class="reference internal" href="_modules/phyloformer/phyloformer.html#load_state_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.phyloformer.load_state_dict" title="Permalink to this definition">#</a></dt>
<dd><p>This function loads a pre-trained phyloformer model to be used for inference
or fine-tuning, from a file containing the state dict. The user must specify the
model’s arrchitecture parameters themselves. For models saved with
<cite>AttentionNet.save</cite> you should use the <cite>load_model</cite> function instead.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model_path</strong> (<em>str</em>) – Path to the saved model (Either the saved state dict or a dictionnary containing
the state dict)</p></li>
<li><p><strong>n_blocks</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of attention blocks, by default 6</p></li>
<li><p><strong>h_dim</strong> (<em>int</em><em>, </em><em>optional</em>) – Hidden dimension of attention blocks, by default 64</p></li>
<li><p><strong>n_heads</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of attentions heads per block, by default 4</p></li>
<li><p><strong>n_seqs</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of sequences in input alignments, by default 20</p></li>
<li><p><strong>seq_len</strong> (<em>int</em><em>, </em><em>optional</em>) – Length of sequences in input alignments, by default 200</p></li>
<li><p><strong>device</strong> (<em>str</em><em>, </em><em>optional</em>) – PyTorch device, by default “cpu”</p></li>
<li><p><strong>dropout</strong> (<em>float</em><em>, </em><em>optional</em>) – Drop out rate, by default 0.0</p></li>
<li><p><strong>single_gpu</strong> (<em>bool</em><em>, </em><em>optional</em>) – Wether inference/fine-tuning will be done on a single GPU, by default True</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A functional instance of AttentionNet ready for use for inference/fine-tuning</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet">AttentionNet</a></p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-phyloformer.pretrained">
<span id="phyloformer-pretrained-module"></span><h2>phyloformer.pretrained module<a class="headerlink" href="#module-phyloformer.pretrained" title="Permalink to this heading">#</a></h2>
<p>The pretrained module contains functions to load functional pre-trainde instances
of the Phyloformer model.</p>
<dl class="py function">
<dt class="sig sig-object py" id="phyloformer.pretrained.evosimz">
<span class="sig-prename descclassname"><span class="pre">phyloformer.pretrained.</span></span><span class="sig-name descname"><span class="pre">evosimz</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet"><span class="pre">AttentionNet</span></a></span></span><a class="reference internal" href="_modules/phyloformer/pretrained.html#evosimz"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.pretrained.evosimz" title="Permalink to this definition">#</a></dt>
<dd><p>Loads a pre-trained Phyloformer model.
This model was trained on Evosimz simulations</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A pre-trained instance of Phyloformer</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet">AttentionNet</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="phyloformer.pretrained.seqgen">
<span class="sig-prename descclassname"><span class="pre">phyloformer.pretrained.</span></span><span class="sig-name descname"><span class="pre">seqgen</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet"><span class="pre">AttentionNet</span></a></span></span><a class="reference internal" href="_modules/phyloformer/pretrained.html#seqgen"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.pretrained.seqgen" title="Permalink to this definition">#</a></dt>
<dd><p>Loads a pre-trained Phyloformer model.
This model was trained on Seq-Gen simulations</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A pre-trained instance of Phyloformer</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet">AttentionNet</a></p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-phyloformer.training">
<span id="phyloformer-training-module"></span><h2>phyloformer.training module<a class="headerlink" href="#module-phyloformer.training" title="Permalink to this heading">#</a></h2>
<p>The training submodule contains necessary functions to train a Phyloformer network</p>
<dl class="py function">
<dt class="sig sig-object py" id="phyloformer.training.init_training">
<span class="sig-prename descclassname"><span class="pre">phyloformer.training.</span></span><span class="sig-name descname"><span class="pre">init_training</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet"><span class="pre">AttentionNet</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">optimizer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Adam'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">learning_rate</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1e-05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loss</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'L2'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">warmup</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">warmup_steps</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">Optimizer</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_LRScheduler</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ReduceLROnPlateau</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Loss</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/phyloformer/training.html#init_training"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.training.init_training" title="Permalink to this definition">#</a></dt>
<dd><p>Creates and returns an optimizer, learning rate scheduler and loss function
from a config dict</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet"><em>AttentionNet</em></a>) – Phyloformer model to train</p></li>
<li><p><strong>optimizer</strong> (<em>str</em><em>, </em><em>optional</em>) – Optimization algorithm (“Adam”, “AdamW”, or “SGD”), by default “Adam”</p></li>
<li><p><strong>learning_rate</strong> (<em>float</em><em>, </em><em>optional</em>) – Learning rate to apply, by default 1e-5</p></li>
<li><p><strong>loss</strong> (<em>str</em><em>, </em><em>optional</em>) – Loss function (choices “L1” or “L2”), by default “L2”</p></li>
<li><p><strong>warmup</strong> (<em>bool</em><em>, </em><em>optional</em>) – Wether to do learning rate warmupt, by default False</p></li>
<li><p><strong>warmup_steps</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of warmup steps, by default 20</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A tuple containing:</dt><dd><ul class="simple">
<li><p>The optimizer</p></li>
<li><p>The learning rate scheduler</p></li>
<li><p>The loss function</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[torch.optim.Optimizer,Scheduler,torch.nn.modules.loss._Loss]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="phyloformer.training.load_checkpoint">
<span class="sig-prename descclassname"><span class="pre">phyloformer.training.</span></span><span class="sig-name descname"><span class="pre">load_checkpoint</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'cpu'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet"><span class="pre">AttentionNet</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Optimizer</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_LRScheduler</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ReduceLROnPlateau</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">_Loss</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/phyloformer/training.html#load_checkpoint"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.training.load_checkpoint" title="Permalink to this definition">#</a></dt>
<dd><p>Loads a model training checkpoint from disk</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<em>str</em>) – Path to the saved checkpoint</p></li>
<li><p><strong>device</strong> (<em>str</em><em>, </em><em>optional</em>) – Device to load the model onto (“cpu” or “cuda”), by default “cpu”</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A tuple containing:</dt><dd><ul class="simple">
<li><p>The model</p></li>
<li><p>The optimizer</p></li>
<li><p>The learning rate scheduler</p></li>
<li><p>The loss function</p></li>
<li><p>The training configuration</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[    <a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet">AttentionNet</a>, torch.optim.Optimizer, Scheduler,     torch.nn.modules.loss._Loss, Dict[str, Any]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="phyloformer.training.save_checkpoint">
<span class="sig-prename descclassname"><span class="pre">phyloformer.training.</span></span><span class="sig-name descname"><span class="pre">save_checkpoint</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet"><span class="pre">AttentionNet</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">optimizer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optimizer</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scheduler</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">_LRScheduler</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ReduceLROnPlateau</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/phyloformer/training.html#save_checkpoint"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.training.save_checkpoint" title="Permalink to this definition">#</a></dt>
<dd><p>Saves a checkpoint of the current training process</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet"><em>AttentionNet</em></a>) – Model to save</p></li>
<li><p><strong>optimizer</strong> (<em>torch.optim.Optimizer</em>) – Optimizer to save</p></li>
<li><p><strong>scheduler</strong> (<em>Scheduler</em>) – Learning rate scheduler to save</p></li>
<li><p><strong>config</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em>) – Config with which training was started</p></li>
<li><p><strong>path</strong> (<em>str</em>) – Path to save the checkpoint to</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="phyloformer.training.training_loop">
<span class="sig-prename descclassname"><span class="pre">phyloformer.training.</span></span><span class="sig-name descname"><span class="pre">training_loop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet"><span class="pre">AttentionNet</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">optimizer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optimizer</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scheduler</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">_LRScheduler</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ReduceLROnPlateau</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">criterion</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">_Loss</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataLoader</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataLoader</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'cpu'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epochs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">80</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">amp</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clip_gradients</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">early_stopping</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stopping_steps</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">best_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">checkpoint_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tensorboard_writer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet"><span class="pre">AttentionNet</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/phyloformer/training.html#training_loop"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#phyloformer.training.training_loop" title="Permalink to this definition">#</a></dt>
<dd><p>Trains a Phyloformer model</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet"><em>AttentionNet</em></a>) – model instance to train</p></li>
<li><p><strong>optimizer</strong> (<em>torch.optim.Optimizer</em>) – Optimizer for training</p></li>
<li><p><strong>scheduler</strong> (<em>Scheduler</em>) – Learning rate scheduler</p></li>
<li><p><strong>criterion</strong> (<em>torch.nn.modules.loss._Loss</em>) – Loss function</p></li>
<li><p><strong>train_data</strong> (<em>DataLoader</em>) – Training data</p></li>
<li><p><strong>val_data</strong> (<em>DataLoader</em>) – Validation data</p></li>
<li><p><strong>device</strong> (<em>str</em><em>, </em><em>optional</em>) – Device on which to train (“cpu” or “cuda”), by default “cpu”</p></li>
<li><p><strong>epochs</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of epochs to train the model for, by default 80</p></li>
<li><p><strong>amp</strong> (<em>bool</em><em>, </em><em>optional</em>) – Use automatic mixed precision (only available on “cuda” device), by default True</p></li>
<li><p><strong>clip_gradients</strong> (<em>bool</em><em>, </em><em>optional</em>) – Wether to clip gradients during training, by default True</p></li>
<li><p><strong>early_stopping</strong> (<em>bool</em><em>, </em><em>optional</em>) – Wether to stop trainig early if there is no improvement of validation loss
, by default False</p></li>
<li><p><strong>stopping_steps</strong> (<em>int</em><em>, </em><em>optional</em>) – After how many stesp without improvement to stop training early, by default 8</p></li>
<li><p><strong>best_path</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Path to save the best checkpoint so fat, by default None</p></li>
<li><p><strong>checkpoint_path</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Path to save the last checkpoint, by default None</p></li>
<li><p><strong>log_file</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Path to the log file to write training metrics, by default None</p></li>
<li><p><strong>tensorboard_writer</strong> (<em>Optional</em><em>[</em><em>Any</em><em>]</em><em>, </em><em>optional</em>) – SummaryWriter for following training on tensorboard, by default None</p></li>
<li><p><strong>config</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em><em>, </em><em>optional</em>) – Config dictionary (saved in checkpoints), by default dict()</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>a tuple containing:</dt><dd><ul class="simple">
<li><p>The model with the best validation loss</p></li>
<li><p>The epoch at which this function returns</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[<a class="reference internal" href="#phyloformer.phyloformer.AttentionNet" title="phyloformer.phyloformer.AttentionNet">AttentionNet</a>, int]</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-phyloformer">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-phyloformer" title="Permalink to this heading">#</a></h2>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          <a class="prev-page" href="modules.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">API reference</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Luca Nesterenko
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">phyloformer package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-phyloformer.attentions">phyloformer.attentions module</a><ul>
<li><a class="reference internal" href="#phyloformer.attentions.KernelAxialMultiAttention"><code class="docutils literal notranslate"><span class="pre">KernelAxialMultiAttention</span></code></a><ul>
<li><a class="reference internal" href="#phyloformer.attentions.KernelAxialMultiAttention.forward"><code class="docutils literal notranslate"><span class="pre">KernelAxialMultiAttention.forward()</span></code></a></li>
<li><a class="reference internal" href="#phyloformer.attentions.KernelAxialMultiAttention.training"><code class="docutils literal notranslate"><span class="pre">KernelAxialMultiAttention.training</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-phyloformer.data">phyloformer.data module</a><ul>
<li><a class="reference internal" href="#phyloformer.data.TensorDataset"><code class="docutils literal notranslate"><span class="pre">TensorDataset</span></code></a></li>
<li><a class="reference internal" href="#phyloformer.data.load_alignment"><code class="docutils literal notranslate"><span class="pre">load_alignment()</span></code></a></li>
<li><a class="reference internal" href="#phyloformer.data.load_dataset"><code class="docutils literal notranslate"><span class="pre">load_dataset()</span></code></a></li>
<li><a class="reference internal" href="#phyloformer.data.load_tree"><code class="docutils literal notranslate"><span class="pre">load_tree()</span></code></a></li>
<li><a class="reference internal" href="#phyloformer.data.write_dm"><code class="docutils literal notranslate"><span class="pre">write_dm()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-phyloformer.phyloformer">phyloformer.phyloformer module</a><ul>
<li><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet"><code class="docutils literal notranslate"><span class="pre">AttentionNet</span></code></a><ul>
<li><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet.forward"><code class="docutils literal notranslate"><span class="pre">AttentionNet.forward()</span></code></a></li>
<li><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet.infer_dm"><code class="docutils literal notranslate"><span class="pre">AttentionNet.infer_dm()</span></code></a></li>
<li><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet.infer_tree"><code class="docutils literal notranslate"><span class="pre">AttentionNet.infer_tree()</span></code></a></li>
<li><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet.save"><code class="docutils literal notranslate"><span class="pre">AttentionNet.save()</span></code></a></li>
<li><a class="reference internal" href="#phyloformer.phyloformer.AttentionNet.training"><code class="docutils literal notranslate"><span class="pre">AttentionNet.training</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#phyloformer.phyloformer.load_model"><code class="docutils literal notranslate"><span class="pre">load_model()</span></code></a></li>
<li><a class="reference internal" href="#phyloformer.phyloformer.load_state_dict"><code class="docutils literal notranslate"><span class="pre">load_state_dict()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-phyloformer.pretrained">phyloformer.pretrained module</a><ul>
<li><a class="reference internal" href="#phyloformer.pretrained.evosimz"><code class="docutils literal notranslate"><span class="pre">evosimz()</span></code></a></li>
<li><a class="reference internal" href="#phyloformer.pretrained.seqgen"><code class="docutils literal notranslate"><span class="pre">seqgen()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-phyloformer.training">phyloformer.training module</a><ul>
<li><a class="reference internal" href="#phyloformer.training.init_training"><code class="docutils literal notranslate"><span class="pre">init_training()</span></code></a></li>
<li><a class="reference internal" href="#phyloformer.training.load_checkpoint"><code class="docutils literal notranslate"><span class="pre">load_checkpoint()</span></code></a></li>
<li><a class="reference internal" href="#phyloformer.training.save_checkpoint"><code class="docutils literal notranslate"><span class="pre">save_checkpoint()</span></code></a></li>
<li><a class="reference internal" href="#phyloformer.training.training_loop"><code class="docutils literal notranslate"><span class="pre">training_loop()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-phyloformer">Module contents</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/scripts/furo.js"></script>
    </body>
</html>