<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="API reference" href="modules.html" /><link rel="prev" title="Welcome to Phyloformerâ€™s documentation!" href="index.html" />

    <!-- Generated with Sphinx 6.1.3 and Furo 2022.12.07 -->
        <title>Introduction to Phyloformer - Phyloformer 0.0.1a3 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?digest=91d0f0d1c444bdcb17a68e833c7a53903343c195" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">Phyloformer 0.0.1a3 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">Phyloformer 0.0.1a3 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Introduction to Phyloformer</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="modules.html">API reference</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="phyloformer.html">phyloformer package</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="phyloformer.scripts.html">phyloformer.scripts package</a></li>
</ul>
</li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="introduction-to-phyloformer">
<h1>Introduction to Phyloformer<a class="headerlink" href="#introduction-to-phyloformer" title="Permalink to this heading">#</a></h1>
<p align="center">
  <img src="https://github.com/lucanest/Phyloformer/blob/main/figures/phyloformer_color.png?raw=true">
</p><section id="phyloformer-towards-fast-and-accurate-phylogeny-estimation-with-self-attention-networks">
<h2>Phyloformer: Towards fast and accurate Phylogeny estimation with self-attention Networks<a class="headerlink" href="#phyloformer-towards-fast-and-accurate-phylogeny-estimation-with-self-attention-networks" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>Luca Nesterenko</p></li>
<li><p>Bastien Boussau</p></li>
<li><p>Laurent Jacob</p></li>
</ul>
<p>This repository contains the scripts for <a class="reference external" href="https://www.biorxiv.org/content/10.1101/2022.06.24.496975v1">the paper</a>:</p>
<div class="highlight-bibtex notranslate"><div class="highlight"><pre><span></span><span class="nc">@article</span><span class="p">{</span><span class="nl">Nesterenko2022phyloformer</span><span class="p">,</span>
<span class="w">  </span><span class="na">author</span><span class="p">=</span><span class="s">{Nesterenko Luca, Boussau Bastien, Jacob Laurent}</span><span class="p">,</span>
<span class="w">  </span><span class="na">title</span><span class="p">=</span><span class="s">{Phyloformer: towards fast and accurate phylogeny estimation with self-attention networks}</span><span class="p">,</span>
<span class="w">  </span><span class="na">year</span><span class="p">=</span><span class="s">{2022}</span><span class="p">,</span>
<span class="w">  </span><span class="na">doi</span><span class="p">=</span><span class="s">{https://doi.org/10.1101/2022.06.24.496975}</span><span class="p">,</span>
<span class="w">  </span><span class="na">url</span><span class="p">=</span><span class="s">{https://www.biorxiv.org/content/10.1101/2022.06.24.496975v1}</span><span class="p">,</span>
<span class="w">  </span><span class="na">journal</span><span class="p">=</span><span class="s">{bioRxiv}</span>
<span class="p">}</span>
</pre></div>
</div>
<a class="reference external image-reference" href="https://github.com/lucanest/Phyloformer/blob/main/figures/sketch.png?raw=true"><img alt="" src="https://github.com/lucanest/Phyloformer/blob/main/figures/sketch.png?raw=true" /></a>
<section id="install">
<h3>Install<a class="headerlink" href="#install" title="Permalink to this heading">#</a></h3>
<p>The easiest way to install the software is using mamba:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">n</span> <span class="n">base</span> <span class="o">-</span><span class="n">c</span> <span class="n">conda</span><span class="o">-</span><span class="n">forge</span> <span class="n">mamba</span>
<span class="n">mamba</span> <span class="n">env</span> <span class="n">create</span> <span class="o">-</span><span class="n">f</span> <span class="n">environment</span><span class="o">.</span><span class="n">yml</span>
<span class="n">conda</span> <span class="n">activate</span> <span class="n">phylo</span>
</pre></div>
</div>
<section id="test-run">
<h4>Test run<a class="headerlink" href="#test-run" title="Permalink to this heading">#</a></h4>
<p>To check that the installation is successful one can run</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">predict</span><span class="o">.</span><span class="n">py</span> <span class="n">testdata</span><span class="o">/</span><span class="n">alignments</span>
</pre></div>
</div>
<p>and then</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">testdata</span><span class="o">/</span><span class="n">test</span><span class="o">.</span><span class="n">py</span>
</pre></div>
</div>
<p>the printed mean normalized Robinson-Fould distance should be equal to 0.063.</p>
</section>
</section>
<section id="usage">
<h3>Usage<a class="headerlink" href="#usage" title="Permalink to this heading">#</a></h3>
<p>We provide the software as a ready to use tool for phylogenetic inference, one can simply run</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">predict</span><span class="o">.</span><span class="n">py</span>
</pre></div>
</div>
<p>providing as argument a directory containing multiple sequence protein alignments in .fasta format,
the program will then predict a phylogenetic tree for each alignment and write them in Newick format in the same directory.</p>
<p>If the user wishes to choose a different directory where the predictions will be saved he can specify it with the <code class="docutils literal notranslate"><span class="pre">--o</span></code> flag.</p>
<p>By default the Phyloformer model used for inference is the one trained on simulations based on the PAM model of evolution, a different model to use can be specified with the <code class="docutils literal notranslate"><span class="pre">--m</span></code> flag (the other one currently available being the one trained on <a class="reference external" href="https://gitlab.com/ztzou/phydl/-/tree/master/evosimz">Evosimz</a> simulations).</p>
<p>Finally, if an NVIDIA gpu is available, <code class="docutils literal notranslate"><span class="pre">--gpu</span> <span class="pre">true</span></code> allows to exploit it offering a great speed up in inference.</p>
</section>
<section id="simulations-and-training">
<h3>Simulations and training<a class="headerlink" href="#simulations-and-training" title="Permalink to this heading">#</a></h3>
<p>To train the network one needs to simulate phylogenetic trees and alignments of sequences evolving along them.</p>
<section id="simulating-the-trees">
<h4>Simulating the trees<a class="headerlink" href="#simulating-the-trees" title="Permalink to this heading">#</a></h4>
<p>The trees can be generated with</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">simulations</span><span class="o">/</span><span class="n">simulateTrees</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">nleaves</span> <span class="o">&lt;</span><span class="n">number</span> <span class="n">of</span> <span class="n">leaves</span> <span class="ow">in</span> <span class="n">each</span> <span class="n">tree</span><span class="o">&gt;</span> <span class="p">(</span><span class="n">default</span> <span class="mi">20</span><span class="p">)</span> <span class="o">--</span><span class="n">ntrees</span> <span class="o">&lt;</span><span class="n">number</span> <span class="n">of</span> <span class="n">trees</span><span class="o">&gt;</span> <span class="o">--</span><span class="nb">type</span> <span class="o">&lt;</span><span class="n">tree</span> <span class="n">topology</span><span class="o">&gt;</span> <span class="p">(</span><span class="n">default</span> <span class="n">uniform</span><span class="p">)</span> <span class="o">--</span><span class="n">o</span> <span class="o">&lt;</span><span class="n">output</span> <span class="n">directory</span><span class="o">&gt;</span> <span class="o">--</span><span class="n">bl</span> <span class="o">&lt;</span><span class="n">branch</span> <span class="n">lenght</span> <span class="n">distribution</span><span class="o">&gt;</span> <span class="p">(</span><span class="n">default</span> <span class="n">uniform</span><span class="p">)</span>
</pre></div>
</div>
<p>The currently supported types of tree topologies are uniform (as in the paper, sampling uniformly from the tree topologies having nleaves), and birth-death (the tree is generated through a birth death process with a birth_rate of 1 and a death_rate of 0.5).</p>
<p>The currently supported types of branch length distribution are uniform (as in the paper, branch lenghts sampled uniformly between 0.002 and 1), and
exponential (branch lenghts sampled from an exponential distribution with a $lambda$ parameter of 0.15)</p>
<p>Therefore to train the network just as in the paper one can create the tree dataset simply with</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">simulations</span><span class="o">/</span><span class="n">simulateTrees</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">ntrees</span> <span class="mi">100000</span> <span class="o">--</span><span class="n">o</span> <span class="o">&lt;</span><span class="n">output</span> <span class="n">directory</span><span class="o">&gt;</span>
</pre></div>
</div>
</section>
<section id="simulating-the-alignments">
<h4>Simulating the alignments<a class="headerlink" href="#simulating-the-alignments" title="Permalink to this heading">#</a></h4>
<p>Currently the supported sequence simulator is <a class="reference external" href="http://tree.bio.ed.ac.uk/software/seqgen/">Seq-Gen</a></p>
<p>The alignments can be generated with</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">simulations</span><span class="o">/</span><span class="n">simulateAlignments</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">i</span> <span class="o">&lt;</span><span class="nb">input</span> <span class="n">directory</span> <span class="k">with</span> <span class="n">the</span> <span class="o">.</span><span class="n">nwk</span> <span class="n">tree</span> <span class="n">files</span><span class="o">&gt;</span>  <span class="o">--</span><span class="n">o</span> <span class="o">&lt;</span><span class="n">output</span> <span class="n">directory</span><span class="o">&gt;</span> <span class="o">--</span><span class="n">l</span> <span class="o">&lt;</span><span class="n">length</span> <span class="n">of</span> <span class="n">the</span> <span class="n">simulated</span> <span class="n">sequences</span><span class="o">&gt;</span> <span class="p">(</span><span class="n">default</span> <span class="mi">200</span><span class="p">)</span> <span class="o">--</span><span class="n">sg</span> <span class="o">&lt;</span><span class="n">path</span> <span class="n">to</span> <span class="n">Seq</span><span class="o">-</span><span class="n">Gen</span><span class="o">-</span><span class="mf">1.3.4</span><span class="o">/</span><span class="n">source</span><span class="o">/&gt;</span> <span class="o">--</span><span class="n">m</span> <span class="o">&lt;</span><span class="n">model</span> <span class="n">of</span> <span class="n">evolution</span><span class="o">&gt;</span> <span class="p">(</span><span class="n">default</span> <span class="n">PAM</span><span class="p">)</span>
</pre></div>
</div>
<p>the possible models of evolution being those supported by Seq-Gen.</p>
<p>Again, to follow the paper one can just do</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">simulations</span><span class="o">/</span><span class="n">simulateAlignments</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">i</span> <span class="o">&lt;</span><span class="nb">input</span> <span class="n">directory</span> <span class="k">with</span> <span class="n">the</span> <span class="o">.</span><span class="n">nwk</span> <span class="n">tree</span> <span class="n">files</span><span class="o">&gt;</span>  <span class="o">--</span><span class="n">o</span> <span class="o">&lt;</span><span class="n">output</span> <span class="n">directory</span><span class="o">&gt;</span> <span class="o">--</span><span class="n">sg</span> <span class="o">&lt;</span><span class="n">path</span> <span class="n">to</span> <span class="n">Seq</span><span class="o">-</span><span class="n">Gen</span><span class="o">-</span><span class="mf">1.3.4</span><span class="o">/</span><span class="n">source</span><span class="o">/&gt;</span>
</pre></div>
</div>
</section>
<section id="creating-a-tensor-dataset">
<h4>Creating a tensor dataset<a class="headerlink" href="#creating-a-tensor-dataset" title="Permalink to this heading">#</a></h4>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">training</span><span class="o">/</span><span class="n">make_tensors</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">treedir</span> <span class="o">&lt;</span><span class="nb">input</span> <span class="n">directory</span> <span class="k">with</span> <span class="n">the</span> <span class="o">.</span><span class="n">nwk</span> <span class="n">tree</span> <span class="n">files</span><span class="o">&gt;</span> <span class="o">--</span><span class="n">alidir</span> <span class="o">&lt;</span><span class="nb">input</span> <span class="n">directory</span> <span class="k">with</span> <span class="n">the</span> <span class="n">corresponding</span> <span class="o">.</span><span class="n">fasta</span> <span class="n">alignment</span> <span class="n">files</span><span class="o">&gt;</span>  <span class="o">--</span><span class="n">o</span> <span class="o">&lt;</span><span class="n">output</span> <span class="n">directory</span><span class="o">&gt;</span>
</pre></div>
</div>
</section>
<section id="training-the-model">
<h4>Training the model<a class="headerlink" href="#training-the-model" title="Permalink to this heading">#</a></h4>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">training</span><span class="o">/</span><span class="n">train</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">i</span> <span class="o">&lt;</span><span class="nb">input</span> <span class="n">directory</span> <span class="k">with</span> <span class="n">the</span> <span class="n">training</span> <span class="n">tensors</span><span class="o">&gt;</span> <span class="o">--</span><span class="n">o</span> <span class="o">&lt;</span><span class="n">output</span> <span class="n">directory</span> <span class="n">where</span> <span class="n">the</span> <span class="n">models</span> <span class="n">will</span> <span class="n">be</span> <span class="n">saved</span><span class="o">&gt;</span>  <span class="o">--</span><span class="n">c</span> <span class="o">&lt;</span><span class="n">json</span> <span class="n">configuration</span> <span class="n">file</span> <span class="k">with</span> <span class="n">hyperparameters</span><span class="o">&gt;</span>  <span class="o">--</span><span class="n">load</span> <span class="p">(</span><span class="n">optional</span><span class="p">)</span> <span class="o">&lt;</span><span class="n">path</span> <span class="n">to</span> <span class="n">model</span> <span class="n">to</span> <span class="n">train</span> <span class="n">further</span><span class="o">&gt;</span>
</pre></div>
</div>
<p>In the training directory one will find an example of configuration file, stdconfig.json, i.e. the one used to train the model in the paper.</p>
</section>
</section>
<section id="reproducibility-of-the-results-in-the-paper">
<h3>Reproducibility of the results in the paper<a class="headerlink" href="#reproducibility-of-the-results-in-the-paper" title="Permalink to this heading">#</a></h3>
<p>The datasets simulated using Seq-Gen are available at:</p>
<ul class="simple">
<li><p>PAM model: <a class="reference external" href="https://plmbox.math.cnrs.fr/f/f5a2ed2667a841cba6f0/">https://plmbox.math.cnrs.fr/f/f5a2ed2667a841cba6f0/</a>.</p></li>
<li><p>WAG model: <a class="reference external" href="https://plmbox.math.cnrs.fr/f/834248a35ba64752a6a4/">https://plmbox.math.cnrs.fr/f/834248a35ba64752a6a4/</a>.</p></li>
</ul>
<p>The simulations under the Evosimz model are available at <a class="reference external" href="https://datadryad.org/stash/dataset/doi%253A10.5061%252Fdryad.rbnzs7h91">https://datadryad.org/stash/dataset/doi%253A10.5061%252Fdryad.rbnzs7h91</a>.</p>
</section>
<section id="disclaimer">
<h3>Disclaimer:<a class="headerlink" href="#disclaimer" title="Permalink to this heading">#</a></h3>
<p>Phyloformer is still at the stage of the proof of concept. In
particular, we observed lower performances:</p>
<ul class="simple">
<li><p>For trees with more than 40 leaves. As discussed in the manuscript
this will likely be fixed after we re-train on trees with larger
evolutionary distances.</p></li>
<li><p>For topologies generated under a birth-death model. The topologies
used in the experiments reported in the manuscript were generated
under the populate function of the ete3 package. We are still
investigating the possible reasons for this discrepancy.</p></li>
</ul>
<p>We are making the trained models available for transparency and would
be very interested to hear about cases where these models
underperform.</p>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="modules.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">API reference</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="index.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Home</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Luca Nesterenko
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Introduction to Phyloformer</a><ul>
<li><a class="reference internal" href="#phyloformer-towards-fast-and-accurate-phylogeny-estimation-with-self-attention-networks">Phyloformer: Towards fast and accurate Phylogeny estimation with self-attention Networks</a><ul>
<li><a class="reference internal" href="#install">Install</a><ul>
<li><a class="reference internal" href="#test-run">Test run</a></li>
</ul>
</li>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#simulations-and-training">Simulations and training</a><ul>
<li><a class="reference internal" href="#simulating-the-trees">Simulating the trees</a></li>
<li><a class="reference internal" href="#simulating-the-alignments">Simulating the alignments</a></li>
<li><a class="reference internal" href="#creating-a-tensor-dataset">Creating a tensor dataset</a></li>
<li><a class="reference internal" href="#training-the-model">Training the model</a></li>
</ul>
</li>
<li><a class="reference internal" href="#reproducibility-of-the-results-in-the-paper">Reproducibility of the results in the paper</a></li>
<li><a class="reference internal" href="#disclaimer">Disclaimer:</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/scripts/furo.js"></script>
    </body>
</html>